### PR TITLE
Update base image to Debian; all components

### PR DESCRIPTION
**Pull request by salrashid123**
PR for issue [#74](https://github.com/GoogleCloudPlatform/cloud-sdk-docker/issues/74):

Update baseline image for debian; install all components by default.  Provide -slim and versioned builds.

I'll temp stage the images [here](https://hub.docker.com/r/salrashid123/cloud-sdk-docker/builds/):
  - salrashid123/cloud-sdk-docker:latest
  - salrashid123/cloud-sdk-docker:latest-slim
  - salrashid123/cloud-sdk-docker:159.0.0
  - salrashid123/cloud-sdk-docker:159.0.0-slim
  - salrashid123/cloud-sdk-docker:legacy-latest

for ref: the dockerhub build rules will look like:

docker_build_settings

### CODE REVIEW

**Comment by ahmetb** [comment]
This looks good. Out of curiosity, what's the size difference between alpine and debian-slim?

Also, is it possible to avoid creating extra folders like /debian_slim and have Docker Hub pick up a custom Dockerfile name, such as Dockerfile.slim ?

**Comment by salrashid123** [comment]
re: dockerfile 
 I dont' see it stated outright but i havne't gotten to work with anything other than the default Dockerfile in the folder
- [https://docs.docker.com/docker-hub/builds/](https://docs.docker.com/docker-hub/builds/)

re image sizes:
(interesting...the debian is larger than legacy)

```
REPOSITORY                           TAG                             IMAGE ID            CREATED             SIZE
cloud-sdk-docker-debian              latest                          b2cac14d712d        23 seconds ago      1.64GB
cloud-sdk-docker-debian-slim         latest                          c84b6cfd8c5c        11 minutes ago      474MB
cloud-sdk-docker-alpine              latest                          86eb5815b68d        13 minutes ago      139MB
cloud-sdk-docker-legacy              latest                          7893fd6e4b84        2 hours ago         1.45GB

```

**Comment by ahmetb** [comment]
@salrashid123 this LGTM. Can you update README.md to add this instead of the current "Supported tags" section:

> - **`google/cloud-sdk:latest`, `google/cloud-sdk:VERSION`:** (large image with additional components pre-installed, Debian-based)
> - **`google/cloud-sdk:slim`, `google/cloud-sdk:VERSION-slim`** (smaller image with no components pre-installed, Debian-based)

and we're good to go. In the future we must look at using `alpine` in the slim tag.

Also in the build screenshot above I suggest making 2 changes:

- don't call it `:latest-slim`; `:slim` is fine.
- don't call it `:latest-legacy`; `:legacy` is fine.

File: Dockerfile [hunk 1]
```diff
@@ -1,15 +1,21 @@
...
+    apt-get update && \
+    apt-get install -y google-cloud-sdk \
+        google-cloud-sdk-app-engine-python \
+        google-cloud-sdk-app-engine-java \
+        google-cloud-sdk-app-engine-go \
+        google-cloud-sdk-datalab \
+        google-cloud-sdk-datastore-emulator \
+        google-cloud-sdk-pubsub-emulator \
+        google-cloud-sdk-bigtable-emulator \
+        kubectl && \
```
Comment (ahmetb) [inline Dockerfile]: FYI: you're doing `apt-get install kubectl` that's probably coming from Debian's repos. That's why gcloud is not able to discover it as installed in `gcloud components list`

File: README.md [hunk 2]
```diff
@@ -56,25 +53,21 @@ instance-1  us-central1-a  n1-standard-1               10.240.0.2   8.34.219.29
...
+By default, all gcloud components are installed on the default image: 
 
-For example, for appengine-python:
-```dockerfile
-FROM google/cloud-sdk
-RUN gcloud components install app-engine-python
-```
+- [https://cloud.google.com/sdk/downloads#apt-get](https://cloud.google.com/sdk/downloads#apt-get)
+
+The debian-slim image (google/cloud-sdk-docker:alpine:159.0.0-slim), contains no additional components but you are welcome to extend the image or supply --build-args during the build state:
```
Comment (ahmetb) [inline README.md]: `google/cloud-sdk-docker:alpine:159.0.0-slim` looks wrong.

Also we don't have `:debian-slim` image. It's just `:slim`.

File: README.md [hunk 3]
```diff
@@ -56,25 +53,21 @@ instance-1  us-central1-a  n1-standard-1               10.240.0.2   8.34.219.29
...
 
-If you need to run additional gcloud components, you need to extend the base image:
+By default, all gcloud components are installed on the default image: 
 
-For example, for appengine-python:
-```dockerfile
-FROM google/cloud-sdk
-RUN gcloud components install app-engine-python
-```
+- [https://cloud.google.com/sdk/downloads#apt-get](https://cloud.google.com/sdk/downloads#apt-get)
```
Comment (ahmetb) [inline README.md]: no need to put this on a separate line.

File: Dockerfile (hunk 1 above)
Comment (salrashid123) [inline Dockerfile]: yes, i filed a bug to allow gcloud components viewer to detect it.  i'll cc you into that, if you want

File: README.md [hunk 4]
```diff
@@ -56,25 +53,21 @@ instance-1  us-central1-a  n1-standard-1               10.240.0.2   8.34.219.29
...
+
+The debian-slim image (google/cloud-sdk-docker:alpine:159.0.0-slim), contains no additional components but you are welcome to extend the image or supply --build-args during the build state:
 
-or for app-engine-java
-```dockerfile
-FROM google/cloud-sdk
-RUN apk --update add openjdk7-jre
-RUN gcloud components install app-engine-java
 ```
+ docker build --build-arg INSTALL_COMPONENTS="google-cloud-sdk-datastore-emulator" -t my-cloud-sdk-docker:159.0.0-slim .
```
Comment (ahmetb) [inline README.md]: you show a build command, but there's no Dockerfile?

File: README.md [hunk 5]
```diff
@@ -56,25 +53,21 @@ instance-1  us-central1-a  n1-standard-1               10.240.0.2   8.34.219.29
...
 
 ### Google App Engine base
 
 The original image in this repository was based off of 
 
 > FROM gcr.io/google_appengine/base
 
-The full Dockerfile for that can be found [here](google_appengine_base/Dockerfile) for archival.
\ No newline at end of file
+The full Dockerfile for that can be found [here](google_appengine_base/Dockerfile) for archival as well as in image tag google/cloud-sdk-docker:latest-legacy
```
Comment (ahmetb) [inline README.md]: call this `:legacy`.

File: alpine/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+ARG SHA256SUM=5b408575407514f99ad913bd0c6991be4b46408ddc7080a6494bbf43e6ce222a
+ENV PATH /google-cloud-sdk/bin:$PATH
+RUN apk --no-cache add curl python py-crcmod bash libc6-compat && \
+    curl -O https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-${CLOUD_SDK_VERSION}-linux-x86_64.tar.gz && \
+    echo "${SHA256SUM}  google-cloud-sdk-${CLOUD_SDK_VERSION}-linux-x86_64.tar.gz" > google-cloud-sdk-${CLOUD_SDK_VERSION}-linux-x86_64.tar.gz.sha256 && \
+    sha256sum -c google-cloud-sdk-${CLOUD_SDK_VERSION}-linux-x86_64.tar.gz.sha256 && \
+    tar xzf google-cloud-sdk-${CLOUD_SDK_VERSION}-linux-x86_64.tar.gz && \
+    rm google-cloud-sdk-${CLOUD_SDK_VERSION}-linux-x86_64.tar.gz google-cloud-sdk-${CLOUD_SDK_VERSION}-linux-x86_64.tar.gz.sha256 && \
+    ln -s /lib /lib64 && \
+    gcloud config set core/disable_usage_reporting true && \
```
Comment (ahmetb) [inline alpine/Dockerfile]: wasn't the decision in the earlier PR to remove this:

     gcloud config set core/disable_usage_reporting true

File: Dockerfile (hunk 1 above)
Comment (ahmetb) [inline Dockerfile]: I don't think you should expect it to detect it if you have not installed it via gcloud. :)

**Comment by ahmetb** [comment]
Mostly LGTM , left some cosmetics comments. It would be great if we can address those.

File: README.md (hunk 2 above)
Comment (salrashid123) [inline README.md]: ok

File: README.md (hunk 5 above)
Comment (salrashid123) [inline README.md]: done

File: alpine/Dockerfile (hunk 6 above)
Comment (salrashid123) [inline alpine/Dockerfile]: we decided we woudnt' enable it by default; its disabled here with that command for clarity

Comment (ahmetb) [inline alpine/Dockerfile]: Got it, thanks.

File: README.md [hunk 7]
```diff
@@ -53,14 +53,13 @@ instance-1  us-central1-a  n1-standard-1               10.240.0.2   8.34.219.29
 
 ### Installing additional components
 
-By default, all gcloud components are installed on the default image: 
+By default, all gcloud components are installed on the default image:  [https://cloud.google.com/sdk/downloads#apt-get](https://cloud.google.com/sdk/downloads#apt-get)
```
Comment (ahmetb) [inline README.md]: FYI no need to use markdown syntax if you're just pasting the link. :) (no need to fix at this time)

File: README.md [hunk 8]
```diff
@@ -56,25 +53,20 @@ instance-1  us-central1-a  n1-standard-1               10.240.0.2   8.34.219.29
...
 
-If you need to run additional gcloud components, you need to extend the base image:
+By default, all gcloud components are installed on the default image:  [https://cloud.google.com/sdk/downloads#apt-get](https://cloud.google.com/sdk/downloads#apt-get)
 
-For example, for appengine-python:
-```dockerfile
-FROM google/cloud-sdk
-RUN gcloud components install app-engine-python
-```
+The :slim image (google/cloud-sdk-docker:alpine:159.0.0-slim), contains no additional components but you are welcome to extend the image or supply --build-args during the build state:
```
Comment (ahmetb) [inline README.md]: > google/cloud-sdk-docker:alpine:159.0.0-slim

still wrong. Maybe just say:

> The `google/cloud-sdk-docker:slim` image contains no ...

**Comment by salrashid123** [comment]
PTAL.

* [build staged images](https://hub.docker.com/r/salrashid123/cloud-sdk-docker/builds/)  (just look at the most recent build set, timewise): 
 * [candidate branch to merge](https://github.com/salrashid123/cloud-sdk-docker-1/tree/update-rebased-image)

I've already update the actual google/cloud-sdk docker hub images as suggested.

The next time anyone pushes to this repo, it'll trigger a build of
:latest
:legacy
:slim

to get tagged versions, someone needs to tag the git repo itself and push
eg. something like
```
export TAG_VER=159.0.0
git tag -a $TAG_VER
git push origin $TAG_VER
```
that will trigger the builds for the named versions.


I also noticed the main dockerhub repo has a webtrigger which will build the legacy,latest,slim automatically (but not named versions; i'll address that in a bit

**Comment by ahmetb** [comment]
Merged and tagged as `159.0.0`.

The `git tag -a $TAG_VER` requires a message. Should we update instructions to say `git tag -a $TAG_VER -m "$TAG_VER"` so that it doesn't prompt? (I just followed this command for now.)

**Comment by icio** [comment]
I spent a few hours after the switch to Alpine getting everything working again. Has that been abandoned? Can we have a tag for it too, please?

**Pull request merged**